### PR TITLE
Backports for 0.3.0-rc4

### PR DESCRIPTION
**Pull request by ax3l**
Backported the following to the `release-0.3.0` branch, updating it to release candidate 4 (rc4):

- `GetIonizerList` fix missing include #2040
- Fix Missing Field Absorber in Z (3D) #2042
- update pizdaint environment #2030
- Examples: Collisional Ionization Include #2037
- fix FillAllGaps #2045
- fix boost 1.64.0 odeint integrate bug #2053
- remove extra typename in documentation of manipulators #2044
- fix shadowed member variable #2051
- Warning in Socket Write #2063
- Fix charge conservation in ThomasFermi #2046
- Abort on CUDA8 + Boost 1.64.0 #2069 

### To Do

- [x] ~~@n01r just found we introduced an other regression with the latest TF PR in #2046, we will wait for that before we merge this PR ->~~ **Update** false alert, RT test was against an outdated PR without the fix

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -103,7 +103,8 @@ for detailed instructions.
  - Support for clang as host && device compiler #1933
 
 **Bug Fixes:**
- - PIC:
+ - PIConGPU:
+   - 3D3V: missing absorber in z #2042
```
Comment (PrometheusPi) [inline CHANGELOG.md]: QUESTION:
why 3V? It occurs in 3D?

File: src/picongpu/include/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp [hunk 2]
```diff
@@ -264,29 +268,31 @@ namespace ionization
...
-
                 /* density in sim units */
-                float_X density = densityV[0];
+                float_X const density = densityV[0];
                 /* energy density in sim units */
-                float_X kinEnergyDensity = kinEnergyV[0];
+                float_X const kinEnergyDensity = kinEnergyV[0];
 
-                /* this is the point where actual ionization takes place */
+                /* Returns the new number of bound electrons for an integer number of macro electrons */
```
Comment (n01r) [inline src/picongpu/include/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp]:
```c++
/* Returns an integer number of new free macro electrons to create */
```

File: src/picongpu/include/simulation_defines/param/ionizer.param [hunk 3]
```diff
@@ -19,7 +19,28 @@
 
 /* @file ionizer.param
  *
- * Ionization Model Configuration
+ * This file contains the proton and neutron numbers of commonly used elements
+ * of the periodic table. The elements here should have a matching list of
+ * ionization energies in @see ionizationEnergies.param. Moreover this file
+ * contains a description of how to configure an ionization model for a species.
+ * Currently each species can only be assigned exactly one ionization model.
```
Comment (n01r) [inline src/picongpu/include/simulation_defines/param/ionizer.param]: > Currently each species can only be assigned exactly one ionization model.

This does not apply anymore. :)

`/* Each species can be assigned a list of ionization models that will be applied sequentially */`

File: src/picongpu/include/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp [hunk 4]
```diff
@@ -41,118 +41,182 @@ namespace particles
...
+                 * state of an ion if necessary, but ignore recombination of
+                 * electrons as prediced by the implemented detailed balance
+                 * algorithm.
+                 */
+                if( prevBoundElectrons > newBoundElectrons )
+                    numNewFreeMacroElectrons = static_cast< uint32_t >( prevBoundElectrons - newBoundElectrons );
             }
 
-            return float_X(0.);
+            return numNewFreeMacroElectrons;
```
Comment (n01r) [inline src/picongpu/include/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp]: After the last change something went wrong again and we get too many electrons. :cry: 

I will look for it

**Comment by ax3l** [comment]
@n01r ok, pls fix the two changes directly in `dev`

**Comment by ax3l** [comment]
thx for the reviews, will update the release branch!

**Pull request merged**